### PR TITLE
Route based permissions

### PR DESCRIPTION
**Pull request by feedmypixel**
Follow up pr to https://github.com/uktrade/data-hub-frontend/pull/1079

This work:

- Will render a `403` at the routing level, rather than hitting the API for a `403`. Uses previous permissions in the frontend application
- Only shows `Interactions` on the dashboard if user has correct permissions
- Renders a `403` if user tries to go to search url without correct permissions, rather than empty results
- Moves the documents permissions to routes

#### Dashboard
lep-dashboard

#### Events search
screen shot 2017-12-27 at 14 39 38

#### Dashboard
da-dashboard

#### Events search
da-events-search

#### Dashboard
dit_user-dashboard

#### Events search
screen shot 2017-12-27 at 14 44 35

### CODE REVIEW

File: test/unit/apps/middleware.test.js [hunk 1]
```diff
@@ -212,4 +212,170 @@ describe('Apps middleware', () => {
...
+      this.middleware.handleRoutePermissions([
+        {
+          path: mockUrl,
+          permissions: [
+            'permission1',
+          ],
+        },
+      ])(this.reqMock, this.resMock, this.nextSpy)
+
+      expect(this.nextSpy).to.be.calledWith()
```
Comment (yeahfro) [inline test/unit/apps/middleware.test.js]: Tempting to consolidate the previous test with this..?

File: test/unit/apps/middleware.test.js [hunk 2]
```diff
@@ -212,4 +212,170 @@ describe('Apps middleware', () => {
...
+      this.middleware.handleRoutePermissions([
+        {
+          path: mockUrl,
+          permissions: [
+            'permission2',
+          ],
+        },
+      ])(this.reqMock, this.resMock, this.nextSpy)
+
+      expect(this.nextSpy).to.be.calledWith({ statusCode: 403 })
```
Comment (yeahfro) [inline test/unit/apps/middleware.test.js]: Tempting to consolidate the previous test with this..?

File: test/unit/apps/middleware.test.js [hunk 3]
```diff
@@ -212,4 +212,170 @@ describe('Apps middleware', () => {
...
+
+    it('should call next with error when route not permitted', () => {
+      this.middleware.handleRoutePermissions([
+        {
+          path: mockUrl,
+          permissions: [],
+        },
+      ])(this.reqMock, this.resMock, this.nextSpy)
+
+      expect(this.nextSpy).to.be.calledWith()
```
Comment (yeahfro) [inline test/unit/apps/middleware.test.js]: Tempting to consolidate the previous test with this..?

File: test/unit/apps/middleware.test.js [hunk 4]
```diff
@@ -212,4 +212,170 @@ describe('Apps middleware', () => {
...
+
+      expect(this.nextSpy).to.be.calledOnce
+    })
+
+    it('should call next with error when route not permitted', () => {
+      this.middleware.handleRoutePermissions([
+        { path: mockUrl },
+      ])(this.reqMock, this.resMock, this.nextSpy)
+
+      expect(this.nextSpy).to.be.calledWith()
```
Comment (yeahfro) [inline test/unit/apps/middleware.test.js]: Tempting to consolidate the previous test with this..?

File: src/apps/companies/constants.js [hunk 5]
```diff
@@ -1,3 +1,8 @@
+const DEFAULT_COLLECTION_QUERY = {
```
Comment (ztolley) [inline src/apps/companies/constants.js]: What does this have to do with permissions?

File: src/apps/middleware.js [hunk 6]
```diff
@@ -20,6 +25,34 @@ function removeBreadcrumb (req, res, next) {
...
+  }), 'permissions')
+
+  if (
+    isUndefined(routePermissions) ||
+    userHasPermission(routePermissions, userPermissions)
+  ) {
+    return true
+  }
+
+  return false
```
Comment (yeahfro) [inline src/apps/middleware.js]: could we just replace this with `return isUndefined(routePermissions) || userHasPermission(routePermissions, userPermissions)` ?

File: src/apps/companies/constants.js (hunk 5 above)
Comment (feedmypixel) [reply src/apps/companies/constants.js]: Please have a look at the commit message. Technically it doesn't but the permissions work has involved moving items into `constants.js` files. So this is a tidy up as I am in this area.

File: src/apps/middleware.js (hunk 6 above)
Comment (feedmypixel) [reply src/apps/middleware.js]: that gave me the eslint error:
screen shot 2017-12-27 at 15 31 55

so have compromised and gone with 
`return isUndefined(routePermissions) || userHasPermission(routePermissions, userPermissions)`

File: test/unit/apps/middleware.test.js (hunk 1 above)
Comment (feedmypixel) [reply test/unit/apps/middleware.test.js]: So after @ztolley and @teneightfive mentioned a test should be doing 1 thing I have started to follow that methodology. I quite like it if you can move things into `beforeEach`. Maybe something we should all discuss?

File: test/unit/apps/middleware.test.js (hunk 2 above)
Comment (feedmypixel) [reply test/unit/apps/middleware.test.js]: see above ☝️

File: test/unit/apps/middleware.test.js (hunk 3 above)
Comment (feedmypixel) [reply test/unit/apps/middleware.test.js]: see above, above☝️

File: test/unit/apps/middleware.test.js (hunk 4 above)
Comment (feedmypixel) [reply test/unit/apps/middleware.test.js]: see above, above, above ☝️ 😁

File: test/unit/apps/middleware.test.js (hunk 1 above)
Comment (ztolley) [reply test/unit/apps/middleware.test.js]: Yup. The pattern we have followed that naturally flows from the use of context is context describes a situation, beforeEach sets it up and makes the call to the function and it is followed by multiple 'it' to test different aspects of the result. 

A test should test 'one thing' which not only keeps the test focussed but also makes it clear exactly what failed when it does it makes it easier to diagnose the cause.
